### PR TITLE
Engine - Develop WazuhDB primitives

### LINKED ISSUE #13519: Engine - Create a set of primitives to interact with WazuhDB

### PR DESCRIPTION
**Pull request by nmkoremblum**
|Related issue|
|---|
|#13519|

## Description

This PR aims to close #13519 by developing a set of primitives to allow the Wazuh Engine to interact with the WazuhDB.

## Logs/Alerts example

<details><summary>Tests Results</summary>

```bash
╭─root@201-dev-engine ~/repos/wazuh/src/engine
╰─# /root/repos/wazuh/src/engine/build/test/source/wdb/wdb_test 
Running main() from /root/repos/wazuh/src/engine/build/_deps/googletest-src/googletest/src/gtest_main.cc
[==========] Running 40 tests from 6 test suites.
[----------] Global test environment set-up.
[----------] 14 tests from wdbTests_SocketInterface
[ RUN      ] wdbTests_SocketInterface.SocketConnectError
[       OK ] wdbTests_SocketInterface.SocketConnectError (0 ms)
[ RUN      ] wdbTests_SocketInterface.SocketConnect
[       OK ] wdbTests_SocketInterface.SocketConnect (0 ms)
[ RUN      ] wdbTests_SocketInterface.SendMessageError
[       OK ] wdbTests_SocketInterface.SendMessageError (0 ms)
[ RUN      ] wdbTests_SocketInterface.SendLongMessageError
[       OK ] wdbTests_SocketInterface.SendLongMessageError (0 ms)
[ RUN      ] wdbTests_SocketInterface.SendEmptyMessageError
[       OK ] wdbTests_SocketInterface.SendEmptyMessageError (0 ms)
[ RUN      ] wdbTests_SocketInterface.SendInvalidSocketError
[       OK ] wdbTests_SocketInterface.SendInvalidSocketError (0 ms)
[ RUN      ] wdbTests_SocketInterface.SendWrongSocketFDError
[       OK ] wdbTests_SocketInterface.SendWrongSocketFDError (0 ms)
[ RUN      ] wdbTests_SocketInterface.SendMessage
[       OK ] wdbTests_SocketInterface.SendMessage (0 ms)
[ RUN      ] wdbTests_SocketInterface.RecvMessage
[       OK ] wdbTests_SocketInterface.RecvMessage (0 ms)
[ RUN      ] wdbTests_SocketInterface.SendRecvMessage
[       OK ] wdbTests_SocketInterface.SendRecvMessage (0 ms)
[ RUN      ] wdbTests_SocketInterface.SendRecvLongestMessage
[       OK ] wdbTests_SocketInterface.SendRecvLongestMessage (0 ms)
[ RUN      ] wdbTests_SocketInterface.remoteCloseBeforeSend
[       OK ] wdbTests_SocketInterface.remoteCloseBeforeSend (0 ms)
[ RUN      ] wdbTests_SocketInterface.remoteCloseBeforeRcv
[       OK ] wdbTests_SocketInterface.remoteCloseBeforeRcv (0 ms)
[ RUN      ] wdbTests_SocketInterface.localSendremoteCloseBeforeRcv
[       OK ] wdbTests_SocketInterface.localSendremoteCloseBeforeRcv (0 ms)
[----------] 14 tests from wdbTests_SocketInterface (1 ms total)

[----------] 5 tests from wdb_connector
[ RUN      ] wdb_connector.Init
[       OK ] wdb_connector.Init (0 ms)
[ RUN      ] wdb_connector.ConnectErrorInexistentSocket
[       OK ] wdb_connector.ConnectErrorInexistentSocket (0 ms)
[ RUN      ] wdb_connector.ConnectErrorNotSocket
[       OK ] wdb_connector.ConnectErrorNotSocket (0 ms)
[ RUN      ] wdb_connector.Connect
[       OK ] wdb_connector.Connect (0 ms)
[ RUN      ] wdb_connector.connectManyTimes
[       OK ] wdb_connector.connectManyTimes (0 ms)
[----------] 5 tests from wdb_connector (0 ms total)

[----------] 5 tests from wdb_query
[ RUN      ] wdb_query.EmptyString
[       OK ] wdb_query.EmptyString (0 ms)
[ RUN      ] wdb_query.TooLongString
[       OK ] wdb_query.TooLongString (0 ms)
[ RUN      ] wdb_query.ConnectAndQuery
[       OK ] wdb_query.ConnectAndQuery (0 ms)
[ RUN      ] wdb_query.SendQueryWithoutConnect
[       OK ] wdb_query.SendQueryWithoutConnect (0 ms)
[ RUN      ] wdb_query.SendQueryConexionClosed
[       OK ] wdb_query.SendQueryConexionClosed (0 ms)
[----------] 5 tests from wdb_query (0 ms total)

[----------] 3 tests from wdb_tryQuery
[ RUN      ] wdb_tryQuery.SendQueryOK_firstAttemp
[       OK ] wdb_tryQuery.SendQueryOK_firstAttemp (0 ms)
[ RUN      ] wdb_tryQuery.SendQueryOK_retry
[       OK ] wdb_tryQuery.SendQueryOK_retry (0 ms)
[ RUN      ] wdb_tryQuery.SendQueryIrrecoverable
[       OK ] wdb_tryQuery.SendQueryIrrecoverable (0 ms)
[----------] 3 tests from wdb_tryQuery (0 ms total)

[----------] 10 tests from wdb_parseResult
[ RUN      ] wdb_parseResult.ParseResultOk
[       OK ] wdb_parseResult.ParseResultOk (0 ms)
[ RUN      ] wdb_parseResult.ParseResultOkWithPayload
[       OK ] wdb_parseResult.ParseResultOkWithPayload (0 ms)
[ RUN      ] wdb_parseResult.ParseResultDue
[       OK ] wdb_parseResult.ParseResultDue (0 ms)
[ RUN      ] wdb_parseResult.ParseResultDueWithPayload
[       OK ] wdb_parseResult.ParseResultDueWithPayload (0 ms)
[ RUN      ] wdb_parseResult.ParseResultError
[       OK ] wdb_parseResult.ParseResultError (0 ms)
[ RUN      ] wdb_parseResult.ParseResultErrorWithPayload
[       OK ] wdb_parseResult.ParseResultErrorWithPayload (0 ms)
[ RUN      ] wdb_parseResult.ParseResultIgnore
[       OK ] wdb_parseResult.ParseResultIgnore (0 ms)
[ RUN      ] wdb_parseResult.ParseResultIgnoreWithPayload
[       OK ] wdb_parseResult.ParseResultIgnoreWithPayload (0 ms)
[ RUN      ] wdb_parseResult.ParseResultUnknown
[       OK ] wdb_parseResult.ParseResultUnknown (0 ms)
[ RUN      ] wdb_parseResult.ParseResultUnknownWithPayload
[       OK ] wdb_parseResult.ParseResultUnknownWithPayload (0 ms)
[----------] 10 tests from wdb_parseResult (0 ms total)

[----------] 3 tests from wdb_tryQueryAndParseResult
[ RUN      ] wdb_tryQueryAndParseResult.SendQueryOK_firstAttemp_wopayload
[       OK ] wdb_tryQueryAndParseResult.SendQueryOK_firstAttemp_wopayload (0 ms)
[ RUN      ] wdb_tryQueryAndParseResult.SendQueryOK_retry_wpayload
[       OK ] wdb_tryQueryAndParseResult.SendQueryOK_retry_wpayload (0 ms)
[ RUN      ] wdb_tryQueryAndParseResult.SendQueryIrrecoverable
[       OK ] wdb_tryQueryAndParseResult.SendQueryIrrecoverable (0 ms)
[----------] 3 tests from wdb_tryQueryAndParseResult (0 ms total)

[----------] Global test environment tear-down
[==========] 40 tests from 6 test suites ran. (4 ms total)
[  PASSED  ] 40 tests.
```

</details>

### CODE REVIEW

File: src/engine/source/wdb/src/unixSocketInterface.cpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+ssize_t recvWaitAll(int sock, void* buf, size_t size) noexcept
+{
+    ssize_t offset {}; // offset in the buffer
+    ssize_t recvb {};  // Recived bytes
+
+    for (offset = 0; offset < size; offset += recvb)
+    {
+        recvb = recv(sock, (char*)buf + offset, size - offset, 0);
+
+        if (0 >= recvb)
```
[the remark this answers is not in the document]
Comment (juliancnn) [reply src/engine/source/wdb/src/unixSocketInterface.cpp:41]: this is due to new code style

File: src/engine/source/main.cpp [hunk 2]
```diff
@@ -27,6 +27,7 @@
 #include <register.hpp>
 #include <router.hpp>
 #include <hlp/hlp.hpp>
+#include <wdb/wdb.hpp>
```
[the remark this answers is not in the document]
Comment (juliancnn) [reply src/engine/source/main.cpp:30]: Yes you're right

File: src/engine/test/source/wdb/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    unixSocketInterface_test.cpp
+    wdb_test.cpp
+)
+
+target_include_directories(wdb_test
+    PRIVATE
+    ${ENGINE_SOURCE_DIR}/wdb/src
+)
+
+# TODO check if it's ok to add base for the json dep
```
Comment (LucioDonda) [inline src/engine/test/source/wdb/CMakeLists.txt:11]: should this comment be erased?

File: src/engine/test/source/wdb/unixSocketInterface_test.cpp [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    // Accept connection
+    const int clientRemote = testAcceptConnection(serverSocketFD);
+    ASSERT_GT(clientRemote, 0);
+
+    // close remote before recv
+    close(clientRemote);
+
+    // gracefully closed
+    ASSERT_THROW(
+        try { recvString(clientLocal); } catch (const RecoverableError& e) {
```
Comment (LucioDonda) [inline src/engine/test/source/wdb/unixSocketInterface_test.cpp:266]: is this line c-lang format compliant?

File: src/engine/test/source/wdb/wdb_test.cpp [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+    auto wdb = WazuhDB(TEST_SOCKET_PATH);
+
+    std::thread t([&]() {
+        const int clientRemote = testAcceptConnection(serverSocketFD);
+        socketinterface::recvString(clientRemote).c_str();
+        socketinterface::sendMsg(clientRemote, TEST_RESPONSE);
+        close(clientRemote);
+    });
+
+    ASSERT_STREQ(wdb.tryQuery(TEST_MESSAGE, 5).c_str(), TEST_RESPONSE);
```
Comment (LucioDonda) [inline src/engine/test/source/wdb/wdb_test.cpp:184]: why 5 attempts, why not leaving it with the default 2 ?

File: src/engine/source/wdb/src/wdb.cpp [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+#include <logging/logging.hpp>
+
+#include "unixSocketInterface.hpp"
+
+namespace wazuhdb
+{
+
+WazuhDB::~WazuhDB()
+{
+    if (this->m_fd > 0)
```
Comment (LucioDonda) [inline src/engine/source/wdb/src/wdb.cpp:15]: we should change all the if to Value Operator Variable in order to leave it as stated in the code style

File: src/engine/source/wdb/src/wdb.cpp [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+void WazuhDB::connect()
+{
+    if (std::filesystem::exists(m_path) == false)
+    {
+        const std::string msg {"The wdb socket does not exist:" + m_path.string()};
+        throw std::runtime_error(msg);
+    }
+    else if (std::filesystem::is_socket(m_path) == false)
+    {
+        const std::string msg {"The wdb socket does not a socket:" + m_path.string()};
```
Comment (LucioDonda) [inline src/engine/source/wdb/src/wdb.cpp:30]: "... is not a socket" or " does not refer to a socket"

File: src/engine/test/source/wdb/CMakeLists.txt (hunk 3 above)
Comment (juliancnn) [reply src/engine/test/source/wdb/CMakeLists.txt:11]: Yes, you are right

File: src/engine/source/wdb/src/wdb.cpp (hunk 7 above)
Comment (juliancnn) [reply src/engine/source/wdb/src/wdb.cpp:30]: Yes

**Review by juliancnn — APPROVED** [review]
LGTM :rock: :rock:

image

**Comment by LucioDonda** [comment]
LGTM!

**Review by LucioDonda — APPROVED** [review]

**Review by JcabreraC — APPROVED** [review]
LGTM !

**Pull request merged**